### PR TITLE
Add new action for PipelineRun - Edit and Run

### PR DESCRIPTION
**Pull request by marniks7**
# Changes

Add new action for PipelineRun - Edit and Run.
Solves: https://github.com/tektoncd/dashboard/issues/2437
* Action from list of PRs
image
* Action from single PR
image
* Result of Edit and Run click:
image
* YAML page loading while PipelineRun is not fetched yet
image

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (new features, significant UI changes, API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Add new feature: ability to edit and run new pipeline (completed or failed or running)
```

### CODE REVIEW

File: src/containers/CreatePipelineRun/YAMLEditor.test.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.test.js:261]: We already have this in at least one other test, can it be extracted to an appropriately named function for reuse?

File: src/containers/PipelineRun/PipelineRun.js
Comment (AlanGreene) [inline src/containers/PipelineRun/PipelineRun.js:342]: nit: `Edit and run`
We use sentence case for UI elements: https://carbondesignsystem.com/guidelines/content/writing-style#use-sentence-case-capitalization

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.js:38]: Could we reuse the existing code from the `rerunPipelineRun` API? Similar to what we've done with `createPipelineRun`, having a function to generate the sanitised payload.

**Review by AlanGreene — CHANGES_REQUESTED** [review]
Thanks for the PR @marniks7. Overall it looks pretty good. I've left a few suggestions, let me know if you have any questions.

As we discussed on one of the previous issues / PRs I'd like to avoid adding more PipelineRun specific code into the YAML editor so we can more easily reuse it for other resource types in the near future.

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (marniks7) [reply src/containers/CreatePipelineRun/YAMLEditor.js:14]: >  The existing PipelineRun specific pieces can be left as they're beyond 
I was thinking about sending it as a code from CreatePipelineRun, but existence of PipelineRun specific pieces stopped me from doing that.
Thanks for clarifying this.

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (marniks7) [reply src/containers/CreatePipelineRun/YAMLEditor.js:38]: what about this one? we are not kinda re-running
```
'dashboard.tekton.dev/rerunOf': name
```
why do we need this one?
```
delete payload.metadata.labels['tekton.dev/pipeline'];
```

Comment (AlanGreene) [reply src/containers/CreatePipelineRun/YAMLEditor.js:38]: `'dashboard.tekton.dev/rerunOf': name`

Yes, we can skip that one 👍 

`delete payload.metadata.labels['tekton.dev/pipeline'];`

We shouldn't set any `tekton.dev/*` labels, they're reserved for use by the Tekton Pipelines controllers, and adding them on the input could potentially cause issues. See https://github.com/tektoncd/pipeline/issues/4366 for some context.

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (marniks7) [reply packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js]: This approach is used for the purpose to be similar to the second pipeline run validation.

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (marniks7) [reply packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js:70]: So, I have tried todo that but failed.

`getByRole(/textbox/).textContent;` returns content without space `apiVersion: tekton.dev/v1beta1kind` and cannot be converted to yaml. `getByRole(/textbox/).innerHTML` requires even more manipulations. So only string validation can be used.  In `CreatePipelineRun.test.js` I have done the validation by using regexp.
However, for e2e we have another issue. Tekton has some default parameters https://tekton.dev/docs/pipelines/install/#customizing-basic-execution-parameters, e.g. `timeout: 1h0m0s` are populated by tekton engine one pipeline run creation. Do we want to compare those fields as well? Not sure. One of the options is to get first pipeline run, remove those fields (and fields in our logic), and compare. 
Another approach would be to use after `Create` with `cy.intercept`. In this case we will be able to use json\yaml and compare, but fields removal is required.
Instead of removal, since we don't know all the fields ahead, and they are subject to change, we can construct new pipelinerun with only known fields.
One more possible option is to leverage some advanced testing features \ frameworks for such goal as  skipping unknown fields. However, I am not aware of such, except json schema validation.

And finally the last option would be to use only some contains assertions.That is what I did.

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (marniks7) [reply packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js]: I don't see how we can use `name` since `generateName` is generated by us and what for...

changed to what you have suggested, but with `*`, like `[title*="${pipelineName}-run"]`

File: src/containers/CreatePipelineRun/YAMLEditor.test.js
Comment (marniks7) [reply src/containers/CreatePipelineRun/YAMLEditor.test.js]: after rework it is not needed

**Comment by marniks7** [comment]
Didn't want to create separate PR for an issue with `pipelinePendingStatus`, in yaml it was `status: !<tag:yaml.org,2002:js/undefined>`.
It is fixed

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (AlanGreene) [reply packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js:70]: > Used only one contains assertions for pipeline name

Yes that's fine, we just want to ensure the content is present and reasonable. Checking the name is enough for now. No need to do a full comparison.

File: src/api/pipelineRuns.test.js
Comment (AlanGreene) [inline src/api/pipelineRuns.test.js:373]: nit: the value of an annotation should be a string

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.js:50]: `useEffect` should be used to run effects when certain values change

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.js:197]: We should display some form of loading state, otherwise the user won't know why the editor is disabled.

**Review by AlanGreene — CHANGES_REQUESTED** [review]
/ok-to-test

Please also squash your commits when done

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (AlanGreene) [reply packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js]: My suggestion was to switch from using `generateName` to `name` so we have full control over the name used for the resource instead of relying on kubernetes to add a randomly generated suffix. That would allow us to query for the specific run we're interested in, instead of using a broader match.

However, as long as `pipelineName` is unique to this test run and we're only creating a single PipelineRun, using `*=` will work, we just need to be careful when modifying this test in future.

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [reply src/containers/CreatePipelineRun/YAMLEditor.js:197]: Yes either some kind of spinner or skeleton until the resource is loaded and ready to display.

Carbon provides a number of elements for this purpose:
- https://v7-react.carbondesignsystem.com/?path=/story/components-loading--default
- https://v7-react.carbondesignsystem.com/?path=/story/components-skeleton-skeletonplaceholder--default&knob-Active%20(active)=true&knob-Classes%20with%20different%20sizes=my--skeleton__placeholder--large&knob-Description%20(description)=Active%20loading%20indicator&knob-Small%20(small)=true

Personally I think the Loading component with some text explaining what it's doing would be best here, but if you have another suggestion I'm willing to consider it.

You can find an example of how we use the Loading component in the LoadingShell while the Dashboard application is loading:
https://github.com/tektoncd/dashboard/blob/896eb37908d12c9f1d2af6d60647c5f9f32e3e83/packages/components/src/components/LoadingShell/LoadingShell.js#L81-L84

File: packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js
Comment (AlanGreene) [inline packages/e2e/cypress/e2e/run/actions-pipelinerun.cy.js:59]: We had problems with flakiness in the 'create' e2e so made some changes there, we'll need to do the same here. Specifically the following 2 changes:

- after clicking Create, ensure navigation has completed
  ```js
  cy.contains('h1', 'PipelineRuns');
  ```
- update query to specify an anchor tag as it was occasionally clicking on a matching span instead which caused the test to fail, also no need for `parent()` this way
  ```js
  cy.contains('a', `${pipelineName}-run`).click();
  ```

File: packages/utils/src/utils/index.js
Comment (AlanGreene) [inline packages/utils/src/utils/index.js:364]: no need to use template here:
```suggestion
  return generateName;
```

File: packages/utils/src/utils/index.js
Comment (AlanGreene) [inline packages/utils/src/utils/index.js:363]: This function is only called if `generateName` is not provided, and a valid `name` cannot end with `-` so this function could be simplified to just:

```js
export function getGenerateNamePrefixForNewRun(name) {
  return `${name}-`;
}
```

Instead of having a new function for this I'd suggest just doing it inline where it's used since it's only used in one place…

File: src/api/pipelineRuns.js
Comment (AlanGreene) [inline src/api/pipelineRuns.js:203]: Let's keep the `...labels` above to handle potentially missing labels, set defaults, etc.

Then here for the rerun, just add the one specific to reruns:

`payload.metadata.labels['dashboard.tekton.dev/rerunOf'] = name;`

This approach keeps clear separation between the two and makes it easier to refactor in future.

File: src/api/pipelineRuns.js
Comment (AlanGreene) [inline src/api/pipelineRuns.js:206]: with the suggestion above this check wouldn't be required as `labels` would always be defined

similar below for annotations

File: src/api/pipelineRuns.js
Comment (AlanGreene) [inline src/api/pipelineRuns.js:176]:
```suggestion
export function generateNewPipelineRunPayload({ pipelineRun, rerun }) {
```

This makes it easier to understand the code when you see a call to this function:

`…generateNewPipelineRunPayload({ pipelineRun, rerun: true })`
instead of
`…generateNewPipelineRunPayload(pipelineRun, true)`

File: src/api/pipelineRuns.test.js
Comment (AlanGreene) [inline src/api/pipelineRuns.test.js:332]: I think it would be better to replace `minimum` and `maximum` in these tests with descriptions of what they're actually doing, e.g. 'with labels, annotations, and params' or whatever

This gives more context in test reports etc. without having to revert to reading the code to understand what the test is doing.

File: src/api/pipelineRuns.test.js
Comment (AlanGreene) [inline src/api/pipelineRuns.test.js:489]: This could be combined with one of the tests above to avoid all the setup duplication, same for the labels test below. Logically it's part of the same processing, preserving only the relevant annotations/labels on the generated payload.

File: src/containers/CreatePipelineRun/CreatePipelineRun.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/CreatePipelineRun.js:61]: good catch, thanks 👍

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.js:44]: nit: please revert this to keep them listed alphabetically, it makes it easier to scan the list

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.js:48]: we only want to do this `if(!loading)` right?

what happens in case of an error loading the PipelineRun?

File: src/containers/CreatePipelineRun/YAMLEditor.test.js
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/YAMLEditor.test.js:227]: these two tests could be combined as they're part of the same piece of functionality, one test for the loading state should be sufficient as it's not particularly complex

**Review by AlanGreene — CHANGES_REQUESTED** [review]
Thanks @marniks7, this is looking pretty good now.

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (marniks7) [reply src/containers/CreatePipelineRun/YAMLEditor.js:48]: yeah, added

- in case of the error code mirror content will be empty, but editable (loading true -> false)
- no information about the error will be displayed to user
image

File: src/api/pipelineRuns.test.js
Comment (marniks7) [reply src/api/pipelineRuns.test.js:332]: - `rerun. generate new pipeline run with spec, metadata - name and namespace`
- `rerun. generate new pipeline run with spec, status, metadata - annotations, labels, uid and resourceVersion`

vs

- `rerun. generate new pipeline run. minimum`
- `rerun. generate new pipeline run. maximum`

Here is the rational behind `maximum`
- nobody would bother to read such long description (with listing words)
- nobody would bother to change such description on changes, which makes description incorrect and more harmful
- `maximum` assumed to be used by anybody making changes there and naming helps to convince others about that (sure, not 100% convincible)

Here is the rational behind `Minimum`
- making sure that this is a minimum input and in the absence of other fields function doesn't fail
- by word `minimum` it prevents others from making unrelated modifications in this test (sure, not 100% convincible)

File: src/api/pipelineRuns.js
Comment (marniks7) [reply src/api/pipelineRuns.js:176]: Modern IDEs are cable of solving such problems without making it more complex (adding 2 `{`, `}` with adding for each function `{`, `}` and `rerun`).
image
Github supports definition display and step into (for some languages)

Comment (AlanGreene) [reply src/api/pipelineRuns.js]: There are other benefits beyond the self-documenting nature of the code, e.g. consistency with the rest of the codebase, and easier refactoring if new params added in future (especially with any optional / default values) as order won't affect call-sites.

File: src/containers/CreatePipelineRun/YAMLEditor.js
Comment (AlanGreene) [reply src/containers/CreatePipelineRun/YAMLEditor.js]: OK that will do for now but we may revisit later to display some error message to the user. They can always refresh the page to retry if needed and it should be rare enough unless they're following an old link that was shared / bookmarked.

File: src/api/pipelineRuns.test.js
Comment (AlanGreene) [reply src/api/pipelineRuns.test.js]: > only few would change such description on changes, which makes description incorrect and more harmful

I agree this could be a problem, I shouldn't have listed the affected fields in my example, I was just trying to give an example that explains what it's doing, e.g. "… with all processed fields", perhaps that is a better description. Also, it is up to the maintainers to ensure that if the test changes the test and its description still match / make sense.

>  prevents others from making unrelated modifications in this test

Not sure about this, also as above it would be the responsibility of the maintainers to identify this and provide feedback if it happens.

I still think we can be clearer than 'minimum' and 'maximum' here but I won't block the PR on this.

Comment (marniks7) [reply src/api/pipelineRuns.test.js]: Changed to
- `edit. generate new pipeline run with all processed fields`
- `edit. generate new pipeline run with minimum possible fields`

Another option would be to switch to more business terms, e.g.
- generate new pipeline run from completed pipeline run
at the same time I don't provide 100% of fields (however I can copy-paste from any completed pipeline run and that would make it very close to that)
- generate new pipeline run from non-created pipeline run with minimum fields

**Review by AlanGreene — APPROVED** [review]
Thanks @marniks7. This will be included in the v0.32 release due in the next few days.
/lgtm

**Pull request merged**
